helper/schema/resource.go

```diff
@@ -69,6 +96,13 @@ type DeleteFunc func(*ResourceData, interface{}) error
 // See Resource documentation.
 type ExistsFunc func(*ResourceData, interface{}) (bool, error)
 
+// See Resource documentation for more details.
+type StateMigrationContext struct {
+	StateSchemaVersion int
+	State              *terraform.InstanceState
+}
+type StateMigrateFunc func(*StateMigrationContext, interface{}) error
```

Comment:
Agreed re: return value.

The context struct used to have the _current_ schema version in it as well, until I realized that it's implicitly known based on the current implementation of the migration function.

I left it as a context struct since it fixes the function signature for implementers across any future changes. I didn't have any specific change in mind though, and I don't feel strongly, so if it feels better to you as a `(int, *terraform.InstanceState, interface{})` I can unwind it.